lib/ddtrace/span.rb

```diff
@@ -39,7 +41,9 @@ def initialize(tracer, name, options = {})
 
       @span_id = Datadog::Utils.next_id()
       @parent_id = options.fetch(:parent_id, 0)
-      @trace_id = options.fetch(:trace_id, @span_id)
+      # use span_id for trace_id if given trace_id is nil or zero
+      @trace_id = options.fetch(:trace_id, 0)
+      @trace_id = @span_id if @trace_id.zero?
```

Comment:
Mmm, no, it could not be a new random number, the only value we can set is span_id IMHO. This is actually to catch someone who would pass a bad header. Then the header *is here* so trace_id is passed, but it's interpreted as zero so in the end the constructor of Span tries to set "0" as the trace_id. In that case what we want is consider this is a standard root span, and abort that distributed tracing attempt. I think it's somewhat related to that commit: https://github.com/DataDog/dd-trace-rb/pull/141#discussion-diff-126014439R167 in the original proposal. That marshalling of IDs into strings then converting them back into ints introduces a risk that "trace_id: 0" appears somewhere. I don't see it in the current code path but it could really happen after some refactoring. So I think the above is slightly more robust, it creates root span if trace_id was not passed as a possibly valid integer. As a side note, there are many cases in which distributed tracing can be broken because we can't control what's coming upstream. We can only try and protect ourselves against bad input.